### PR TITLE
Getting ready to submit

### PR DESCRIPTION
**Pull request by nsheff**
Here we are collecting a to-do list and any comments regarding our pending submission to bioconductor, which we hope to do soon (within a few days). Comments are invited!

# To do for submission
- [x] Conform to style guide (line length, spaces after commas)
- [x] Fix test formatting
- [x] pass R CMD Build
- [x] pass R CMD Check on Windows with no errors/warnings
- [x] pass R CMD Check on Linux
- [x] pass R BiocCheck
- [x] Write README (Follow [LOLA](https://github.com/nsheff/LOLA) example)
- [x] Divide MIRA.R file
- [x] See if `extractSampleName` can be replaced with `tools::file_path_sans_ext()`
- [x] Make repo public and add Travis-CI (see #4)
- [x] Solve '"self_contained"' error reported below.
- [x] Solve example problem reported below (https://github.com/databio/MIRA/blob/9c6cd6255b82c0c362c89403057f9ddf2100b943/R/MIRA.R#L717)
- [x] Remove bioc instructions from README (it's not a bioc package yet)
- [x] plot functions should use `theme_classic` and other improvements suggested earlier

### CODE REVIEW

**Comment by nsheff** [comment]
@j-lawson, I enabled travis ( https://travis-ci.org/databio/MIRA) so it just needs the yaml setup -- @vpnagraj can help make sure it's working correctly.

**Comment by nsheff** [comment]
@j-lawson web page is live (http://databio.org/mira/); also, please remove bioconductor instructions on readme (we can add that later)

File: vignettes/ExampleOfUse.Rmd [hunk 1]
```diff
@@ -11,68 +11,84 @@ vignette: >
...
-input: a few RRBS samples from Ewing sarcoma and myeloid cell type (DNA methylation)
-input: regionset that I get from Nathan (EWS specific),myeloid specific TF
+Ewing sarcoma is a pediatric cancer characterized by the fusion of the EWS and FLI1 genes which results in a fusion transcription factor. We want to see if Ewing samples can be distinguished from muscle-related samples based on their respective MIRA scores. To accomplish that, we can compare MIRA scores for Ewing and non-Ewing samples using a set of Ewing-specific DNase hypersensitivity sites (DHSs). A set of muscle-specific DHSs will also be used for comparison.
+input: single-nucleotide resolution DNA methylation data (from reduced representation bisulfite sequencing) for samples from Ewing sarcoma and a collection of muscle related samples including human skeletal muscle myoblasts, myotubes differentiated from a skeletal muscle myoblasts cell line, primary myoblasts from patients with Facioscapulohumeral Muscular Dystrophy (FSHD), and myotubes from FSHD myoblasts.
```
Comment (vreuter) [inline vignettes/ExampleOfUse.Rmd]: Do input data need to be from RRBS, or can they also come from WGBS?

File: R/MIRA.R [hunk 2]
```diff
@@ -56,93 +57,95 @@ if(getRversion() >= "2.15.1"){
...
+        GRDTList = GRList #this case is okay
     }else{
         stop("GRList should be a GRangesList or a list of data.tables")
     }
     
-    if (sampleNameInBSDT){
-        if(!"sampleName" %in% colnames(BSDT)){
-            stop("BSDT should have sampleName col if sampleNameInBSDT=TRUE")
+    if (sampleNameInBSDT) {
+        if (!"sampleName" %in% colnames(BSDT)) {
```
Comment (vreuter) [inline R/MIRA.R]: Is the operator binding such that `if (!a %in% b)` always evaluates the same as `if (!(a %in% b))`? Regardless, I think that it's slightly clearer with parentheses, but just preference.

**Comment by vreuter** [comment]
It looks like a lot of whitespace is changing with respect to passing keyword arguments to named parameters. It's not enforced, but Bioconductor notes a preference for no whitespace around `=` for keyword args. Is there a reason to move away from that preference?

**Comment by vpnagraj** [comment]
the .travis.yml needs to be added to the .Rbuildignore ... it is resulting in a `devtools::check()` note: 

```
Status: 1 NOTE
checking for hidden files and directories ... NOTE
Found the following hidden files and directories:
  .travis.yml
These were most likely included in error. See section ‘Package
structure’ in the ‘Writing R Extensions’ manual.
```

File: R/MIRA.R [hunk 3]
```diff
@@ -16,20 +16,21 @@
...
 #' @importFrom data.table ":=" setDT data.table setkey fread setnames 
-#'             setcolorder melt setkeyv rbindlist setattr
+#'             setcolorder melt setkeyv rbindlist setattr setorder
 #' @importFrom Biobase sampleNames
 #' @importFrom bsseq getBSseq hasBeenSmoothed
 NULL
 
 
-# Because of some issues with CRAN submission,
+# Because of some issues with CRAN submission,
```
Comment (vpnagraj) [inline R/MIRA.R]: not really clear what this code is doing / why it's necessary ... i didn't see any related NOTES from check ... if you do need it, maybe it's better to put in a utility function (not exported) so that's it at least documented

**Comment by vpnagraj** [comment]
the built-in demo data could use some more descriptive help files ... as it is now the documentation seems kind of minimal (`?exampleBins`)

number of rows, number of columns, description of features, units, etc. would be helpful (see for example `?mtcars`)

**Comment by vpnagraj** [comment]
i know this on your list of things to do but just wanted to make a note here of the `BiocCheck::BiocCheck()` results:

```
$error
[1] "At least 80% of man pages documenting exported objects must have runnable examples. The following pages do not:"

$warning
[1] "y of x.y.z version should be even in release"

$note
[1] "Consider shorter lines; 33 lines (1%) are > 80 characters long."              
[2] "Consider indenting lines with a multiple of 4 spaces; 101 lines (4%) are not."
```

File: R/MIRA.R (hunk 2 above)
Comment (nsheff) [inline R/MIRA.R]: Agreed; @j-lawson we should update all such syntax with an extra set of parens.

File: vignettes/ExampleOfUse.Rmd (hunk 1 above)
Comment (nsheff) [inline vignettes/ExampleOfUse.Rmd]: either will work. RRBS is the lower denominator due to sparsity.

File: R/MIRA.R (hunk 3 above)
Comment (nsheff) [inline R/MIRA.R]: what exactly are you referring to? The section on `if(getRversion() >= "2.15.1")`? Unfortunately this is required to pass CMD check due to data.table quirks. See [this stackoverflow question](http://stackoverflow.com/questions/9439256/) for explanation.

**Comment by nsheff** [comment]
you can add a travis badge to the readme.

**Comment by nsheff** [comment]
Suggested by @vpnagraj:

>. I typically use the `pkg::fun` format inside my packages: http://r-pkgs.had.co.nz/namespace.html#imports

I agree. Right now we're doing:
```
#' @import BiocGenerics S4Vectors IRanges

#' @importFrom GenomicRanges GRanges GRangesList elementMetadata strand seqnames 
#'             granges
#' @importFrom ggplot2 ggplot aes facet_wrap geom_boxplot geom_jitter geom_line

#' @importFrom data.table ":=" setDT data.table setkey fread setnames 
#'             setcolorder melt setkeyv rbindlist setattr
#' @importFrom Biobase sampleNames
#' @importFrom bsseq getBSseq hasBeenSmoothed
```

Let's reduce this: 
Keep the `@import` ones; for `@importFrom` I would suggest keeping only data.table, GenomicRanges, and IRanges, and let's reduce the number in data.table (definitely take out `melt` which causes issues). In other words, wherever we use `melt` we should fully qualify it `data.table::melt` and not import; same is probably true for some others (use your discretion) in data.table, but leave `:=`.

File: vignettes/MIRAvignette.Rmd [hunk 4]
```diff
@@ -15,12 +15,15 @@ Note: This vignette gives a broad overview of the MIRA package and some main fun
...
-```{r,echo=FALSE}
+```{r, echo = FALSE}
 library(MIRA)
 data(exampleBins)
 plotMIRARegions(exampleBins)
-exScores=exampleBins[,.(score = scoreDip(methyl,binCount=11)),by=.(featureID,sampleName)]
-exScores[,sampleType := c("Condition1","Condition2")][] #normally would do join with annoDF
+exScores = exampleBins[, .(score = scoreDip(methyl, binCount = 11)), 
+                       by = .(featureID, sampleName)]
+#normally would do join with annoDF
```
Comment (nsheff) [inline vignettes/MIRAvignette.Rmd]: 1. What are these comments for? What's annoDF?
2. I think comments should have spaces after `#`

**Comment by nsheff** [comment]
Ok, getting there. Some comments on vignette 1: 
- you need to guide the user through the vignette. From the very beginning of Vignette 1, overview is fine, but then it just starts abruptly. Introduce each code snippet and each figure; precede it with exactly what this code is about to do, in one sentence.
- generally let's try to have the vignettes have more code and less text; as a general rule, the code should explain what it's doing. ideally you don't want to explain the code exactly in text...
- markdown requires a space after `##` for headings. View your final vignette both in HTML output, but also in github's rendering of the markdown to catch these errors.
- this vignette starts at the end (with a MIRA profile on finished data). Personally I would start at the beginning (loading data)
- the vignette should show all commands the user will need to run. Right now you are hiding important things with `echo = FALSE`
- along these lines, you need to provide a way for a user to easily get data from common formats into the input formats for MIRA; and the intro vignette should document/demonstrate this (in other words, don't use R objects with data already created; show how this data is actually imported. the user will have to do that).
- we should switch from `hitCount` and `readCount` to something more intuitive.
- code or variables should always be encapsulated with ticks so they are displayed as such for clarity.
- we have several warnings. these should be eliminated: (no reason to have warnings in a vignette) `## Warning in FUN(X[[i]], ...): GRList should be a named list/GRangesList. 
##                 Sequential names given according to order in object.`
- function name `scoreDip` should be changed -- and it should be hidden from the user anyway.
- binning can be de-emphasized (I'm not sure the user even needs to be aware of this at all... or at least not much)

I've started to make a few of these changes in my commits.

File: vignettes/ExampleOfUse.Rmd [hunk 5]
```diff
@@ -9,70 +9,106 @@ vignette: >
...
+controlFiles=paste0(controlFiles, rep(1:3, each = 4), ".bed")
+RRBSFiles = c(caseFiles, controlFiles)
+regionSetFiles = c("sknmc_specific.bed", "muscle-specific.bed")
+BSDTList = lapply(X = RRBSFiles, FUN = BSreadBiSeq)
+regionSets = lapply(X = regionSetFiles, FUN = fread)
+regionSets = lapply(X = regionSets, 
+                    FUN = function(x) setNames(x, c("chr", "start", "end")))
+#converting list items from data.table to GRanges objects
+#strand is not relevant for our region sets
+regionSets = lapply(X = regionSets, FUN =
```
Comment (vreuter) [inline vignettes/ExampleOfUse.Rmd]: It looks odd to have the argument on a separate line from the parameter. The way it is in the previous assignment (lines 57-58) looks better.

**Comment by vreuter** [comment]
Again, not enforced, but...
It seems like CRAN/Bioconductor prefer ` <- ` for all assignments. 

https://bioconductor.org/developers/how-to/coding-style/
http://stackoverflow.com/questions/1741820/assignment-operators-in-r-and

Personally, I think that ` <- ` for function definition and ` = ` for other assignments is clearer, but is there a reason to not adopt the community preference?

**Comment by nsheff** [comment]
Yes, `<-` is hideous and not used in any other programming language... :)

**Comment by vreuter** [comment]
Ha OK fair enough. I'd actually prefer all ` = ` to all ` <- ` but just wanted to put it out there that in so doing I guess we're being R renegades? :skull_and_crossbones: :boat: :crossed_swords:

**Comment by nsheff** [comment]
I find the 2nd vignette quite hard to follow. Here are some initial points, we can discuss further:
- make it either evaluated or unevaluated; no reason to do halfway I think...
- must provide an archive of data you use. this vignette should load the data directly from this archive.  the point is that someone else could run this vignette from scratch.
- list formatting isn't rendering correctly; , view your final vignette in BiocStyle HTML output, and in github's rendering of the markdown to catch these errors.
- the vignette should actually show what you're doing, not just describe a list (what you have now in 0. ... 1.1 ... etc.). This is the point of a vignette. Why are we entering at step 3 of the process?
- don't built vignette code around weird file names; fix the file names/folder structure so the vignette is straightforward.
- The vignette is teaching a user what to do, not describing a methods section ("The resize function from the Genomic Ranges package was used to make...").

File: DESCRIPTION [hunk 6]
```diff
@@ -1,7 +1,7 @@
 Package: MIRA
 Version: 0.99.0
-Date: 2016-09-12
-Title: Methylation-based inference of regulatory activity
+Date: 2016-03-30
+Title: Methylation-Based Inference of Regulatory Activity
```
Comment (vreuter) [inline DESCRIPTION]: `Based` --> `based`

File: DESCRIPTION [hunk 7]
```diff
@@ -1,7 +1,7 @@
 Package: MIRA
 Version: 0.99.0
-Date: 2016-09-12
-Title: Methylation-based inference of regulatory activity
+Date: 2016-03-30
```
Comment (vreuter) [inline DESCRIPTION]: Should this actually be Friday 3/31?

**Comment by j-lawson** [comment]
To do list was completed and most other comments were implemented. I still need to finish revisions to the vignettes.
